Added discovery result representation

tomhoefer opened the pull request:
Currently discovery results contain only labels as human readable information to inform about new  things that were discovered. However solutions based on ESH might require an id property to give the solution´s UI a hint how to represent the unique id of the discovered thing. For this reason the discovery result was extended by a representation object containing the unique id and its type like IP address, MAC address, serial number, ...

Signed-off-by: Thomas Höfer <t.hoefer@telekom.de>

tomhoefer commented:
If it is ok for you then I would like to change the HueLightDiscoveryService in a separate PR because I want to check if we can retrieve the lights mac address with a modified jue library...

maggu2810 commented:
I do not use the discovery part of ESH until now.
If I understand your PR correctly that adds at least one field that adds some information to the user.
Shouldn't we add the option, that the bundle / addon that generates the discovery result could add as much additional information as possible?
E.g. for a printer the discovery result could contain printer name, IP address, ... another binding could set a serial number, a port, ...
The UI could display as much information as possible, so the user could evaluate all collected (discovered) info.

Is this already possible or is the intention of this PR another one=

tomhoefer commented:
As far as I know you can already inject properties into the discovery result. Here you could add your necessary data... In contrast to this the new representation object can be considered as an effective key for the user...

kaikreuzer commented on bundles/config/org.eclipse.smarthome.config.discovery.test/src/test/groovy/org/eclipse/smarthome/config/setup/test/inbox/DynamicThingUpdateOSGITest.groovy:
```diff
@@ -44,12 +44,13 @@ import org.junit.Test
     def DEFAULT_TTL = 60
 
-        final BINDING_ID = 'dnamicUpdateBindingId'
+    final BINDING_ID = 'dnamicUpdateBindingId'
```
typo (I see that it was not you, but please fix it anyhow :-)): should be "dynamic"

kaikreuzer commented on bundles/config/org.eclipse.smarthome.config.discovery.test/src/test/groovy/org/eclipse/smarthome/config/setup/test/inbox/DynamicThingUpdateOSGITest.groovy:
```diff
@@ -44,12 +44,13 @@ import org.junit.Test
 
-        final BINDING_ID = 'dnamicUpdateBindingId'
+    final BINDING_ID = 'dnamicUpdateBindingId'
     final THING_TYPE_ID = 'dnamicUpdateThingType'
```
dito

tomhoefer commented:
@dnobel and @kaikreuzer 
I have updated this PR according to our discussions. I have removed the representation object and introduced the getRepresentationProperty operation on the DiscoveryResult interface

kaikreuzer commented:
Thanks, @tomhoefer, looks good to me!
We should just see to soon add documentation for the discovery stuff - it gets more and more complicated and it should be an exception to merge this PR without according documentation.

@maggu2810: As @tomhoefer said, you can already now add any kind of property to a discovered thing. This new feature now (after the latest changes) simply points to one of the properties to say which one should be used as an identifier in the UI for the user. For some devices, this might be the ip address, for others a serial number or mac address, so the UI needs a hint about which property to show. And showing all of them imho would not make sense - it would be a flat and ugly list and most of it not relevant at the time of discovery.
